### PR TITLE
Use`eq_dyn`, `neq_dyn`, `lt_dyn`, `lt_eq_dyn`, `gt_dyn`, `gt_eq_dyn` kernels from arrow

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

~Marking as draft because the PR fails as the Arrow kernels don't handle timestamps and dates, which DataFusion does. I will make a PR to fix arrow shortly~

re https://github.com/apache/arrow-datafusion/issues/1474 -- and towards  #87 

 # Rationale for this change

@Jimexist added dyn kernels in https://github.com/apache/arrow-rs/pull/858 but we still have the dispatch logic in datafusion, which is redundant (and as we add more logic to arrow-rs we will need to change datafusion to take advantage of it)

@matthewmturner  is working on a similar set of kernels for scalars in https://github.com/apache/arrow-rs/pull/1074

In this way, we will have a single set of kernels that datafusion can call and we will extend them as needed

# What changes are included in this PR?
Use different arrow kernels

# Are there any user-facing changes?
No

Corresponding arrow PR: https://github.com/apache/arrow-rs/issues/1086

### CODE REVIEW

**Comment by liukun4515** [comment]
@alamb This pr https://github.com/apache/arrow-datafusion/pull/1483 may be ready next week. 
After comparison the decimal pull request, I will file the pull request about the mathematics operation with decimal type.
I think we can finish the feature of binary operation with decimal first.

**Comment by alamb** [comment]
> @alamb This pr #1483 may be ready next week.
> After comparison the decimal pull request, I will file the pull request about the mathematics operation with decimal type.
> I think we can finish the feature of binary operation with decimal first.

👍  sounds like a great plan to me.

**Comment by liukun4515** [comment]
> > @alamb This pr #1483 may be ready next week.
> > After comparison the decimal pull request, I will file the pull request about the mathematics operation with decimal type.
> > I think we can finish the feature of binary operation with decimal first.
> 
> 👍 sounds like a great plan to me.

thanks.

**Comment by alamb** [comment]
@liukun4515  I am hopefuly that eventually the `dyn_eq` / `dyn_eq_lit` type kernels that @matthewmturner  is working on in https://github.com/apache/arrow-rs/pull/1074 will also support `Decimal` values (though we will have to work out precision and scale)

**Comment by liukun4515** [comment]
> @liukun4515 I am hopefuly that eventually the `dyn_eq` / `dyn_eq_lit` type kernels that @matthewmturner is working on in [apache/arrow-rs#1074](https://github.com/apache/arrow-rs/pull/1074) will also support `Decimal` values (though we will have to work out precision and scale)

The https://github.com/apache/arrow-datafusion/pull/1483 is ready for merged, now I am developing the kernel for decimal operation in the arrow-rs. I think it will ready in the next release of arrow-rs.

**Comment by alamb** [comment]
Rebased on top of https://github.com/apache/arrow-datafusion/pull/1523 (aka upgrade to arrow 7.0.0) which includes  https://github.com/apache/arrow-rs/pull/1095  from @viirya  ❤️  and this PR is in much better shape.

**Comment by alamb** [comment]
This PR is failing because eq_dyn kernels don't have support for Decimal. I will add that

**Comment by alamb** [comment]
Filed https://github.com/apache/arrow-rs/issues/1200 to track adding support for decimal arrays

**Comment by alamb** [comment]
Switching back to draft until the underlying arrow kernel support is done

**Comment by alamb** [comment]
cc @matthewmturner  @viirya  and @liukun4515  I think this is now ready for review

File: datafusion/src/physical_plan/expressions/binary.rs [hunk 1]
```diff
@@ -91,8 +131,10 @@ fn is_not_distinct_from_bool(
         .collect())
 }
 
-// TODO add iter for decimal array
-// TODO move this to arrow-rs
+// TODO move decimal kernels to to arrow-rs
```
Comment (liukun4515) [inline datafusion/src/physical_plan/expressions/binary.rs:134]: 👍

Comment (liukun4515) [reply datafusion/src/physical_plan/expressions/binary.rs:134]: Maybe I can take this, and migrate decimal to arrow-rs kernel recently

**Review by liukun4515 — APPROVED** [review]
LGTM
@alamb codebase is becoming more cleaner.

**Comment by alamb** [comment]
> @alamb codebase is becoming more cleaner.

Thanks -- I am trying @liukun4515 . What I really hope to do with these kernels is get to a point where we can add more specialized implementations in arrow-rs and have them flow automatically to DataFusion.

A particular interest of ours in IOx is performance of dictionary encoded data -- so after this PR is merged, the work that @viirya  is doing in PRs like https://github.com/apache/arrow-rs/pull/1326 will end up automatically being available in DataFusion)

**Review by viirya — COMMENTED** [review]
lgtm

File: datafusion/src/physical_plan/expressions/binary.rs [hunk 2]
```diff
@@ -1194,12 +1236,12 @@ impl BinaryExpr {
...
-            Operator::Gt => binary_array_op!(left, right, gt),
-            Operator::GtEq => binary_array_op!(left, right, gt_eq),
-            Operator::Eq => binary_array_op!(left, right, eq),
-            Operator::NotEq => binary_array_op!(left, right, neq),
+            Operator::Lt => lt_dyn(&left, &right),
+            Operator::LtEq => lt_eq_dyn(&left, &right),
+            Operator::Gt => gt_dyn(&left, &right),
+            Operator::GtEq => gt_eq_dyn(&left, &right),
+            Operator::Eq => eq_dyn(&left, &right),
+            Operator::NotEq => neq_dyn(&left, &right),
```
Comment (viirya) [inline datafusion/src/physical_plan/expressions/binary.rs:1244]: 👍

**Pull request merged**